### PR TITLE
feat(egctl): provide default services

### LINKED ISSUE #1122: Allow the user to not have to specify all K8s resources in `egctl x translate --to=xds`
*Description*:
Today the user must specify all K8s resources when translating from `gateway-api` to `xds`. Support an argument such as `--add-missing-resources` where the user only specifies the Gateway API resources and `egctl` adds the rest (Namespaces, Services, Secrets) as dummy resources.

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

### PR DESCRIPTION
**Pull request by muyuan0**
Fix #1122

### CODE REVIEW

File: internal/cmd/egctl/translate.go [hunk 1]
```diff
@@ -644,6 +735,46 @@ func kubernetesYAMLToResources(str string, addMissingResources bool) (*gatewayap
 				resources.Namespaces = append(resources.Namespaces, namespace)
 			}
 		}
+
+		requiredServiceMap := map[string]map[string]*v1.Service{}
```
Comment (arkodg) [inline internal/cmd/egctl/translate.go:739]: thoughts on using 2 state entities (existingServiceMap, resources.Services)
existingServiceMap := map[string]bool where the key is the namespacedName of the service

So the logic would be - if a backend doesnt exist in map, add it to the map as well as to `resources`

Comment (muyuan0) [reply internal/cmd/egctl/translate.go]: Thanks for your suggestion. It indeed simplifies the logic. There is a need to merge multiple ports of the service in the same name:
```
			inserted := false
			for _, port := range service.Spec.Ports {
				if port.Protocol == servicePort.Protocol && port.Port == servicePort.Port {
					inserted = true
					break
				}
			}

			if !inserted {
				service.Spec.Ports = append(service.Spec.Ports, servicePort)
			}
```
So we still need to track the v1.Service.

Comment (arkodg) [reply internal/cmd/egctl/translate.go]: good catch, agree this is needed

File: internal/cmd/egctl/translate.go [hunk 2]
```diff
@@ -644,6 +730,35 @@ func kubernetesYAMLToResources(str string, addMissingResources bool) (*gatewayap
...
+			addMissingServices(requiredServiceMap, route)
+		}
+
+		providedServiceMap := map[string]bool{}
+		for _, service := range resources.Services {
+			providedServiceMap[service.Namespace+"/"+service.Name] = true
+		}
+
+		for key, service := range requiredServiceMap {
+			if _, found := providedServiceMap[key]; !found {
```
Comment (arkodg) [inline internal/cmd/egctl/translate.go:757]: this logic is not quite right, you are building out an accurate `requiredServiceMap` and inserting missing service as well as service ports, but in this line you are not adding the service if the service exists (but not checking if the ports exist)
what are you thoughts on this suggestion
1. pass an existing `serviceMap` (prepopulated with `resources.Services`) to `addMissingServices`
2. rebuild/rewrite `resources.Services` based on `serviceMap`

Comment (muyuan0) [reply internal/cmd/egctl/translate.go:757]: I would like to rewrite the providedServiceMap (changing the value from bool to v1.Service first), and then rebuild the `resources.Services`. Therefore we don't need to add more arguments to addMissingServices.
What do you think about it?

Comment (muyuan0) [reply internal/cmd/egctl/translate.go:757]: I would like to rewrite the providedServiceMap (changing the value from bool to v1.Service first), and then rebuild the `resources.Services`. Therefore we don't need to add more arguments to addMissingServices.
What do you think about it?

Comment (arkodg) [reply internal/cmd/egctl/translate.go:757]: yeah that works, you would need to prepopulate `providedServiceMap` before  with existing `resources.Services`

**Review by arkodg — APPROVED** [review]
LGTM, thanks !

**Review by AliceProxy — APPROVED** [review]

**Pull request merged**